third_party/amd/lib/TritonAMDGPUToLLVM/ConvertWarpPipeline.cpp

```diff
@@ -0,0 +1,317 @@
+      return rewriter.notifyMatchFailure(forOp, "no pipelined_for");
+    forOp->removeAttr("triton.warp_pipeline.pipelined_for");
+
+    // Look up allocation info as in original pass.
+    auto func = forOp->getParentOfType<mlir::triton::FuncOp>();
+    Allocation *allocation = moduleAllocation.getFuncData(func);
+    if (!allocation)
+      return rewriter.notifyMatchFailure(forOp, "no Allocation for function");
+
+    if (failed(emitPipelinedFor(rewriter, forOp.getLoc(), forOp, allocation)))
+      return failure();
+
+    return success();
+  }
+
+private:
+  LogicalResult emitPipelinedFor(PatternRewriter &builder, Location loc,
+                                 scf::ForOp forOp,
+                                 Allocation *allocation) const {
+    // 1. Insert conditional branch first,
+    builder.setInsertionPointAfter(forOp);
+    // Set barrier before starting the loop. This resolves any outstanding
+    // synchronization before beginning the specialized asymmetric
+    // synchronization.
+    auto preBarrier = builder.create<gpu::BarrierOp>(loc);
+    preBarrier->moveBefore(forOp);
+    builder.setInsertionPointAfter(preBarrier);
+
+    // Insert condbarrier::second_half before starting the loop
+    // FIXME : correctly calculate numbers per the arch
+    auto i32ty = builder.getIntegerType(32);
+    auto workIDX = builder.create<ROCDL::ThreadIdXOp>(loc, i32ty);
+    auto constZero = builder.create<arith::ConstantIntOp>(loc, 0, 32);
+    auto constWarpSize = builder.create<arith::ConstantIntOp>(loc, 256, 32);
+    auto warpIDX = builder.create<arith::DivSIOp>(loc, workIDX, constWarpSize);
+    auto warpLow = builder.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                                 warpIDX, constZero);
+    auto warpHigh = builder.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne,
+                                                  warpIDX, constZero);
+
+    builder.create<mlir::triton::amdgpu::CondBarrierOp>(loc, warpHigh);
+
+    // Insert condbarrier::first_half after the end of the loop
+    builder.setInsertionPointAfter(forOp);
+    builder.create<mlir::triton::amdgpu::CondBarrierOp>(loc, warpLow);
+
+    // 2. Collect existing barrier information.
+    SmallVector<Block *> clusterBlocks;
+    SmallVector<Operation *> clusterOps;
+    SmallVector<bool> bars;
+    std::map<int, Operation *> existingBarrierMap;
+    Operation *terminatorOp = nullptr;
+
+    for (auto &op : *forOp.getBody()) {
+      if (auto exeOp = dyn_cast<scf::ExecuteRegionOp>(op)) {
+        // fail conversion with executeRegion from unkown source.
+        if (exeOp->getAttr("triton.warp_pipeline.stage") == nullptr)
+          return failure();
+        exeOp.setNoInline(false);
+        clusterOps.push_back(&op);
+        clusterBlocks.push_back(&exeOp->getRegion(0).front());
+        bars.push_back(false);
+      } else if (isa<ROCDL::BarrierOp, ROCDL::SBarrierOp,
```

Comment:
Don't think we use `rocdl.barrier` in triton at all? We are missing `gpu.barrier` here which can exist too?